README.md
# LionRockJS MVC
mvc is the core module for LionRockJS

## Controller
`Controller` is the main class for handling requests. It use `state` to store data and handle actions.

## ControllerMixin

A `ControllerMixin` is a class designed to extend the functionality of a `Controller` by allowing you to inject custom logic at various points in the controller's request-processing lifecycle. It provides a way to organize and reuse code that might be common across multiple controllers or actions, operating on the shared `state` object.

In essence, `ControllerMixin`s provide a plugin-like architecture for `Controller`s in LionRockJS, allowing developers to compose controller behavior from smaller, reusable pieces.

## State
The `state` is a core JavaScript Map object. It acts as a central repository for all runtime information related to a single request being processed by a controller instance.

In summary: The `state` object in `Controller` is a dynamic, per-request data store. It encapsulates all relevant information for processing a request and generating a response, serving as a shared context for the `controller` and its extensions (`ControllerMixins`).

### Controller State Keys

The `Controller.state` map uses the following predefined keys to store and manage request and response data:

#### General & Flow Control State Keys

*   `Controller.STATE_CLIENT ('client')`: Stores the controller instance itself.
*   `Controller.STATE_ACTION ('action')`: The name of the action method to be executed (e.g., 'index', 'show').
*   `Controller.STATE_FULL_ACTION_NAME ('fullActionName')`: The fully qualified action name, typically prefixed (e.g., 'action_index').
*   `Controller.STATE_EXITED ('exited')`: A boolean flag. If `true`, it indicates that the controller's execution flow has been intentionally stopped before completion.
*   `Controller.STATE_CHECKPOINT ('checkpoint')`: A general-purpose key that can be used by applications for various purposes, such as pagination tokens or flow control markers.

#### Request-Related State Keys

*   `Controller.STATE_REQUEST ('request')`: Contains the original, raw request object received by the server.
*   `Controller.STATE_REQUEST_BODY ('requestBody')`: Stores the parsed body content from the incoming HTTP request.
*   `Controller.STATE_REQUEST_HEADERS ('requestHeader')`: An object or map of headers received with the incoming HTTP request.
*   `Controller.STATE_REQUEST_COOKIES ('requestCookie')`: An object or map of cookies received with the incoming HTTP request.
*   `Controller.STATE_HOSTNAME ('hostname')`: The hostname that the incoming request was made to.
*   `Controller.STATE_QUERY ('query')`: An object or map of parsed query string parameters from the request URL.
*   `Controller.STATE_PARAMS ('params')`: Parameters extracted from the URL path (e.g., from dynamic route segments) or other request sources.
*   `Controller.STATE_CLIENT_IP ('clientIP')`: The IP address of the client that initiated the HTTP request.
*   `Controller.STATE_USER_AGENT ('userAgent')`: The User-Agent string sent by the client's browser or HTTP client.
*   `Controller.STATE_LANGUAGE ('language')`: Represents the language preference for the current request, often derived from request parameters or headers (e.g., 'en', 'fr').

#### Response-Related State Keys

*   `Controller.STATE_BODY ('body')`: Holds the content that will form the body of the HTTP response.
*   `Controller.STATE_HEADERS ('headers')`: An object or map of headers to be included in the outgoing HTTP response.
*   `Controller.STATE_COOKIES ('cookies')`: An object or map of cookies to be set in the outgoing HTTP response.
*   `Controller.STATE_STATUS ('status')`: The HTTP status code for the outgoing response (e.g., 200 for OK, 404 for Not Found).

## View

The `View` component in LionRockJS MVC is responsible for presenting data. It's designed to take data (usually provided by a `Controller`) and a template file, and then render them into a final output format, which could be HTML, JSON, XML, or plain text.

### Key Features

1.  **Constructor (`constructor(file="", data={}, defaultFile="")`)**:
    *   It's initialized with:
        *   `file`: A string, typically representing the path to a template file (e.g., an HTML template).
        *   `data`: An object containing the data that needs to be rendered within the template.
        *   `defaultFile`: An optional string, possibly a fallback template file if the primary `file` isn't found or specified.

2.  **Static Factory (`static factory(file, data = {})`)**:
    *   This method provides a convenient way to create `View` instances.
    *   It uses `this.DefaultViewClass` to instantiate the view. This allows developers to easily substitute a custom view class for the default one by changing the `View.DefaultViewClass` static property.

3.  **Render Method (`async render()`)**:
    *   This is the core method responsible for producing the output.
    *   In the base `View` class, the `render()` method simply returns `this.data`. This basic implementation indicates that the base `View` class is intended to be extended.
    *   Subclasses would override the `render()` method to implement specific templating logic (e.g., using a templating engine like EJS, Handlebars, Pug, or a custom rendering mechanism) to combine the `this.file` (template) with `this.data` to produce the final output (e.g., an HTML string).

### In Essence

The `View` class provides a standardized interface for rendering. Controllers typically prepare data and then pass it to a `View` instance (or a subclass of `View`) along with a template file path. The `View` instance's `render()` method is then called to generate the response content, which the controller subsequently sends back to the client.

This design allows for different view rendering strategies to be plugged in by creating new classes that inherit from this base `View` and implement their own `render()` logic.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.1.0 (2023-06-08)


### Features

* Controller expose properties to state, update tests ([20199ee](https://github.com/LionRockJS/mvc/commit/20199eeb8c35c10bfbbe0004929abaf4e5dace0c))
* first commit, identical function to @kohanajs/core-mvc 3.2.2 ([e83e17a](https://github.com/LionRockJS/mvc/commit/e83e17a338560196ceefdbe68b6d1ac853aed6ca))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


classes/Controller.mjs
/**
 * Copyright (c) 2023 Kojin Nakana
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

export default class Controller {
  static STATE_CLIENT = 'client';
  static STATE_REQUEST = 'request';
  static STATE_LANGUAGE = 'language';
  static STATE_CLIENT_IP = 'clientIP';
  static STATE_HOSTNAME = 'hostname';
  static STATE_CHECKPOINT = 'checkpoint';
  static STATE_ACTION = 'action';
  static STATE_QUERY = 'query';
  static STATE_PARAMS = 'params';
  static STATE_FULL_ACTION_NAME = 'full_action_name';
  static STATE_HEADERS = 'headers';

  /**
   *
   * @type {ControllerMixin[]}
   */
  static mixins = [];
  static suppressActionNotFound = false;

  #headerSent = false;

  // properties
  error = null;

  body = '';

  headers = {
    "X-Content-Type-Options": "nosniff"
  };

  /**
   *
   * @type {{name: String, value: String, options: {secure:Boolean, maxAge:Number}}[]} cookies
   */
  cookies = [];

  status = 200;

  state = new Map();

  /**
   *
   * @param {Request} request
   */
  constructor(request) {
    const query = request.query || {};
    const params = request.params || {};
    const raw = request.raw || {};

    this.state.set(Controller.STATE_QUERY, query);
    this.state.set(Controller.STATE_PARAMS, params);
    this.state.set(Controller.STATE_CLIENT, this);
    this.state.set(Controller.STATE_REQUEST, request);
    this.state.set(Controller.STATE_LANGUAGE, params.language || query.language);
    this.state.set(Controller.STATE_CLIENT_IP, (!request?.headers) ? '0.0.0.0' : (
      request.headers['cf-connecting-ip']
      || request.headers['x-real-ip']
      || request.headers['x-forwarded-for']
      || request.headers.remote_addr
      || request.ip
    ));
    this.state.set(Controller.STATE_HOSTNAME, raw.hostname);
    this.state.set(Controller.STATE_CHECKPOINT, query.checkpoint || query.cp || null);
    this.state.set(Controller.STATE_ACTION, params.action);
    this.state.set(Controller.STATE_HEADERS, this.headers);

    this.constructor.mixins.forEach(mixin => mixin.init(this.state));
  }

  /**
   *
   * @param {string | null} actionName
   * @returns {object}
   */
  async execute(actionName = null) {
    try {
      // guard check function action_* exist
      const action = `action_${actionName || this.state.get(Controller.STATE_ACTION) || 'index'}`;
      this.state.set(Controller.STATE_FULL_ACTION_NAME, action);
      if (this[action] === undefined) await this.#handleActionNotFound(action);

      // stage 0 : setup
      if (!this.#headerSent) await this.#mixinsSetup();

      // stage 1 : before
      if (!this.#headerSent) await this.#mixinsBefore();
      if (!this.#headerSent) await this.before();

      // stage 2 : action
      if (!this.#headerSent) await this.mixinsAction(action);
      if (!this.#headerSent) await this[action]();

      // stage 3 : after
      if (!this.#headerSent) await this.#mixinsAfter();
      if (!this.#headerSent) await this.after();
    } catch (err) {
      await this.#serverError(err);
    }

    return {
      status: this.status,
      body: this.body,
      headers: this.headers,
      cookies: this.cookies,
    };
  }

  async #handleActionNotFound(action) {
    if (this.constructor.suppressActionNotFound) {
      this[action] = async () => {/***/};
      return;
    }

    await this.#notFound(`${this.constructor.name}::${action} not found`);
  }

  /**
   * @async
   * @callback MixinCallback
   * @param {ControllerMixin} mixin
   */
  /**
   *
   * @param {MixinCallback} lambda
   * @returns {Promise<void>}
   */
  async #loopMixins(lambda) {
    const { mixins } = this.constructor;
    for (let i = 0; i < mixins.length; i++) {
      if (this.#headerSent) break;
      // eslint-disable-next-line no-await-in-loop
      await lambda(mixins[i]);
    }
  }

  async #mixinsSetup() {
    await this.#loopMixins(async mixin => mixin.setup(this.state));
  }

  async #mixinsBefore() {
    await this.#loopMixins(async mixin => mixin.before(this.state));
  }

  async before() {/***/}

  async mixinsAction(fullActionName) {
    await this.#loopMixins(async mixin => mixin.execute(fullActionName, this.state));
  }

  async action_index() {/***/}

  async #mixinsAfter() {
    await this.#loopMixins(async mixin => mixin.after(this.state));
  }

  async after() {/***/}

  /**
   *
   * @param {string} msg
   */
  async #notFound(msg) {
    this.body = `404 / ${msg}`;
    await this.exit(404);
  }

  /**
   *
   * @param {Error} err
   */
  async #serverError(err) {
    this.error = err;
    if (!this.body) this.body = err.message;
    await this.exit(500);
  }

  /**
   *
   * @param {string} location
   * @param {boolean} keepQueryString
   */
  async redirect(location, keepQueryString= false) {
    if(!keepQueryString){
      this.headers.location = location;
    }else{
      const query = new URLSearchParams(this.state.get(Controller.STATE_QUERY));
      const qs = query.toString();
      const delimiter = /\?/.test(location) ? '&' : '?';
      this.headers.location = qs ? `${location}${delimiter}${qs}` : location;
    }
    await this.exit(302);
  }

  /**
   *
   * @param {string} msg
   */
  async forbidden(msg = '') {
    this.body = `403 / ${msg}`;
    await this.exit(403);
  }

  /**
   *
   * @param {Number} code
   */
  async exit(code) {
    this.status = code;
    this.#headerSent = true;

    //exit all mixins
    const { mixins } = this.constructor;
    await Promise.all(mixins.map(async mixin => mixin.exit(this.state)));

    //run delegate onExit
    await this.onExit();
  }

  async onExit(){}
}

Object.freeze(Controller.prototype);